### ISSUE TITLE
Dynamically switch between Emterpreter and WASM

### ISSUE THREAD

**Issue by AjayP13** [emscripten-core/emscripten#7917]
I have made a pretty full-featured Python-in-the-browser library / environment using Emscripten, I am planning on releasing soon:

https://github.com/plasticityai/coldbrew/

It allows you to run Python synchronously (blocking the JavaScript event loop) or asynchronously (yielding back to the JavaScript event loop every N Python byte code instructions). 

Currently, I have to mark all functions on the code path through the CPython Python interpreter as Emterpreter functions. I understand this is a necessity for the "asynchronous" running of Python so that I can yield back to the JavaScript event loop with "emscripten_sleep". However, since Emterpreter functions run slower, it is making the "synchronous" execution of Python also slow, even though the "synchronous" execution doesn't need to call "emscripten_sleep" ever.

What I would like is a way to sometimes run a function in the Emterpreter and sometimes run it without the Emterpreter dynamically at run time (I realize this would store two copies of its instructions, once in WASM, and once in Emterpreter bytecode). 

I have tried to work around this myself by duplicating the function code path using an LLVM IR pass, trying to hack `tools/emterpretify.py` to let me do this with no luck. The CPython code base is fairly complex and it is difficult to just duplicate function symbols to create a secondary code path to mark as Emterpretable. Is there any way to do this with Emscripten or any suggestions on how to proceed / where to modify the code base?

**Comment by curiousdannii** [comment]
I don't think there's any way to do this currently. You could just build it twice, but the two versions wouldn't be able to interact with each other.

Is your main concern that synchronous mode locks the browser? If so, I'd suggest that you make web workers your main use case, and say people should only use Coldbrew in the main JS process if they're sure it won't take much time to run.

**Comment by AjayP13** [comment]
Hm, yes, I have though about building it twice, but the problem is they don't share the same heap memory / state. 

Yes, that is the main concern. I could run it in a web worker, but from what I understand it would limit the ability of the Python script to easily access variables and functions in the scope of `window` without messaging back and forth, which I am trying to avoid as it would also require re-working significant portions of the library.

**Comment by curiousdannii** [comment]
That's true. But directly accessing arbitrary properties of `window` seems to me to be something that, while perhaps useful while prototyping, would not be a good approach for production code. It's quite similar to `eval`, and mostly an anti-pattern IMO.

Up to you what you choose to do of course. But from my experience I'd avoid Emterpreter unless it's actually needed: sync C functions that must be implemented with async JS.

**Comment by kripken** [comment]
Hmm, it may be possible to make them share the heap memory and global
state. With asm.js you can use --separate-asm to emit a .asm.js file, and
with wasm that's in a separate file anyhow (but you need to avoid using
memory init there). Then you can create another asm.js/wasm instance that
looks at the same memory buffer (and for wasm, must be the same table too).
So if one was compiled with the emterpreter and one without, you could swap
between them. There is an option for this, SWAPPABLE_ASM_MODULE, see
settings.js.

This would rely on the two separate builds being identical aside from one
being emterpreted. I think that should work, since static memory and other
things are deterministic, and the emterpreter should not add static memory
- but if we do add the emterpreter data statically, that would need to be
worked around.

**Comment by gabrielcuvillier** [comment]
An alternative solution would be to run your interpreter in a web worker, where blocking is not an issue. 
But maybe it would be a lot of rework

**Comment by AjayP13** [comment]
That's fair, accessing `window` would be an anti-pattern. I warn about the implication of that being similar to `eval()` in the README, but I still would like to have that ability / I think the web worker re-work would be significant, but I might try that if I can't get it working another way. I also don't think the web worker actually provides much speed up as I would still have to use the Emterpreter in the web worker, I believe, since Coldbrew allows the Python code to make HTTP requests or call asynchronous JavaScript routines which would still not be possible in a web worker without the Emterpreter yielding back to web worker's JavaScript event loop.

I think I've actually almost got it working with a little hack on `emterpretify`, will post back here with results in case this thread is useful to anyone else.

**Comment by curiousdannii** [comment]
Wow, clever approach!
